storage/schema.org.md

```diff
@@ -0,0 +1,230 @@
+
+```
+POST /files/actions HTTP/2
+accept: application/ld+json
+content-type: multipart/related; boundary=foo_bar_baz; type=application/ld+json
+
+--foo_bar_baz
+content-type: application/ld+json
+
+{
+    "@context": "http://schema.org",
+    "@type": "CreateAction",
+    "agent": "http://example.com/users/67890",
+    "result": {
+        "@type": "MediaObject",
+        "name": "image.jpg",
+        "alternateName": "My file",
+        "isPartOf": "http://example.com/articles/12345"
+    }
+}
+
+--foo_bar_baz
+content-type: image/jpeg
+
+[DATA]
+--foo_bar_baz
+```
+
+`alternateName` wasn't exposed as an input type, but we can still include it (hello custom properties).
+
+Assuming the action isn't immediately completed, the response is:
+
+```
+HTTP/2 202
+content-type: application/ld+json
+
+{
+    "@context": "http://schema.org",
+    "@type": "CreateAction",
+    "@id": "http://example.com/files/action/1234567890",
+    "actionStatus": "ActiveActionStatus",
+    "startTime": "2019-08-21T14:34:47Z",
+    "agent": "http://example.com/users/67890",
+    "result": {
+        "@type": "MediaObject",
+        "name": "image.jpg",
+        "alternateName": "My file",
+        "isPartOf": "http://example.com/articles/12345"
+    },
+    "potentialAction": {
+        "@type": "CancelAction",
+        "target": {
+            "@type": "EntryPoint",
+            "httpMethod": "POST",
+            "urlTemplate": "/files/actions",
+            "encodingType": "application/json+ld",
+            "contentType": "application/json+ld"
+        },
+        "agent-input": "required"
+        "object": "http://example.com/files/action/1234567890"
+    }
+}
+```
+
+(A separate [`CancelAction`](https://schema.org/CancelAction) can be used to stop it from completing. Might not be technical possible of course, but might make sense for scheduled actions.)
+
+---
+
+Request the `CreateAction` again to see if it's completed:
+
+```
+GET /files/action/1234567890 HTTP/2
+accept: application/ld+json
+```
+
+```
+HTTP/2 200
+content-type: application/ld+json
+
+{
+    "@context": "http://schema.org",
+    "@type": "CreateAction",
+    "@id": "http://example.com/files/action/1234567890",
+    "actionStatus": "CompletedActionStatus",
+    "startTime": "2019-08-21T14:34:47Z",
+    "endTime": "2019-08-21T14:34:49Z",
+    "agent": "http://example.com/users/67890",
+    "result": "http://example.com/files/12345"
+}
+```
+
+It has, there's a new resource at http://example.com/files/12345.
+
+---
+
+Request the new resource:
+
+```
+GET /files/12345 HTTP/2
+accept: application/ld+json
+```
+
+```
+HTTP/2 200
+content-type: application/ld+json
+
+{
+    "@context": "http://schema.org",
+    "@type": "ImageObject",
+    "@id": "http://example.com/files/12345",
+    "name": "image.jpg",
+    "alternateName": "My file",
+    "isPartOf": "http://example.com/articles/12345",
+    "contentSize": 12345,
+    "contentUrl": "http://example.com/public/files/12345.jpg",
```

Comment:
how to model the public link vs something internal? Assuming there is some kind of authentication over this API, this would be the public link then as the API itself would have to give out the private one.

Reply:
Seem to be a few ways. Could be multiple values using [`LinkRole`](https://schema.org/LinkRole), `linkRelationship` is the equivalent of `rel`. [`hasDigitalDocumentPermission`](https://schema.org/hasDigitalDocumentPermission) could provide resource-level access control (could combine the two if you consider the `ImageObject` to have two encodings, one private and one public). Sci.pe also use an [`AudienceRole` on actions](https://sci.pe/get-started/api#audiences) wihich seems related.

Reply:
Should content be advertised if the client requesting if they do not have access to it? Or will this be the case that the client may have access but could not pass it on?

Reply:
It would be good to only show the client what they could do (so can be public). Making clients act on behalf of the end user could well be too complex, but reduces public access (ie I wouldn't be able to see actions that I could still perform through the UI).